File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    def diameterOfBinaryTree(self, root):
        self.res = 0
        self.height(root)
        return self.res

    def height(self, root):

        if root is None:
            return 0

        l_h = self.height(root.left)
        r_h = self.height(root.right)

        self.res = max(self.res, l_h + r_h)

        return max(l_h, r_h) + 1
        

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        height(root);
        return res;
    }

    private int height(TreeNode root){
        if(root == null) return 0;

        int left_h = height(root.left);
        int right_h = height(root.right);

        res = Math.max(res, left_h + right_h);

        return Math.max(left_h, right_h) + 1;
    }
}